clients/src/main/java/org/apache/kafka/clients/consumer/internals/MemberAssignmentReconciler.java

```diff
@@ -0,0 +1,291 @@
+
+        // For assignment, we modify the "current" assignment by adding all entries that were present in the
+        // "target" assignment but not in the "current" assignment.
+        AssignmentModifier modifier = Set::addAll;
+
+        return reconcile(assignment,
+                timer,
+                Operation.ASSIGN,
+                partitionDiffer,
+                AssignPartitionsEvent::new,
+                modifier);
+    }
+
+    private ReconciliationResult reconcile(Optional<Assignment> assignment,
+                                           Timer timer,
+                                           Operation operation,
+                                           PartitionDiffer partitionDiffer,
+                                           EventGenerator eventGenerator,
+                                           AssignmentModifier assignmentModifier) {
+        // Check for any outstanding operations first. If a conclusive result has already been reached, return that
+        // before processing any further.
+        Optional<ReconciliationResult> inflightStatus = checkInflightStatus(timer);
+
+        if (inflightStatus.isPresent())
+            return inflightStatus.get();
+
+        // "diff" the two sets of partitions: our "current" assignment and the "target" assignment. The result is
+        // sorted primarily so when the partitions show up in the logs, it's easier for us humans to understand.
+        Set<TopicPartition> diffPartitions = sortPartitions(partitionDiffer.diff(subscriptions.assignedPartitions(),
+                targetPartitions(metadata, assignment)));
+
+        // TODO: should this be here? Do we expect the above to do anything besides move things around in memory?
+        timer.update();
+
+        if (diffPartitions.isEmpty()) {
+            log.debug("No partitions were {} in the new assignment", operation.verbPastTense);
+            return ReconciliationResult.NO_CHANGE;
+        }
+
+        // Set up our callback invocation. We don't block here waiting on it its completion, though.
+        log.debug("Preparing to invoke {} with the following partitions: {}", operation.methodName, diffPartitions);
+        RebalanceCallbackEvent event = eventGenerator.generate(Collections.unmodifiableList(new ArrayList<>(diffPartitions)));
+        inflightCallback = Optional.of(event);
+
+        // Enqueue it in our background->application shared queue. This should be invoked in the Consumer.poll() method.
+        backgroundEventQueue.add(event);
+
+        // When the core event is completed in the application via Consumer.poll(), we want to invoke our code to
+        // "commit" the change.
+        event.future().whenComplete((result, error) -> {
+            if (error != null) {
+                log.warn("An error occurred when invoking {} with the following partitions: {}", operation.methodName, diffPartitions, error);
+                // TODO: should we proceed or abort?
```

Comment:
The old consumer completely fails (abort the poll loop) with a KafkaException when there is a failure in any of the callbacks (exception starting [here](https://github.com/apache/kafka/blob/9e5ca8416de6011fa8c0d63430e9b89fe751ee2b/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java#L485))

Reply:
This is tricky because it doesn't fail immediately. It waits until _after_ the assignment step. So now we need to keep track of that exception across multiple heartbeats and application<->background thread hops 😢